hera_cal/redcal.py

```diff
@@ -789,3 +837,335 @@ def is_redundantly_calibratable(antpos, bl_error_tol=1.0):
+
+    Returns a dictionary of results with the following keywords:
+        'g_firstcal': firstcal gains in dictionary keyed by ant-pol tuples like (1,'Jxx').
+            Gains are Ntimes x Nfreqs gains but fully described by a per-antenna delay.
+        'gf_firstcal': firstcal gain flags in the same format as 'g_firstcal'. Will be all False.
+        'g_omnical': full omnical gain dictionary (which include firstcal gains) in the same format.
+            Flagged gains will be 1.0s.
+        'gf_omnical': omnical flag dictionary in the same format. Flags arise from NaNs in log/omnical.
+        'v_omnical': omnical visibility solutions dictionary with baseline-pol tuple keys that are the
+            first elements in each of the sub-lists of reds. Flagged visibilities will be 0.0s.
+        'vf_omnical': omnical visibility flag dictionary in the same format. Flags arise from NaNs.
+        'chisq': chi^2 per degree of freedom for the omnical solution. Normalized using noise derived
+            from autocorrelations. If the inferred pol_mode from reds (see redcal.parse_pol_mode) is
+            '1pol' or '2pol', this is a dictionary mapping antenna polarization (e.g. 'Jxx') to chi^2.
+            Otherwise, there is a single chisq (because polarizations mix) and this is a numpy array.
+        'chisq_per_ant': dictionary mapping ant-pol tuples like (1,'Jxx') to the average chisq
+            for all visibilities that an antenna participates in.
+        'omni_meta': dictionary of information about the omnical convergence and chi^2 of the solution
+    '''  
+    rv = {}  # dictionary of return values
+    rc = RedundantCalibrator(reds)
+    if freqs is None:
+        freqs = data.freqs
+    if times_by_bl is None:
+        times_by_bl = data.times_by_bl
+    
+    # perform firstcal
+    d_fc = rc.firstcal(data, df=np.median(np.ediff1d(freqs)))
+    d_fc_rd = rc.remove_degen_gains(d_fc)
+    rv['g_firstcal'] = {ant: np.array(np.exp(2j * np.pi * np.outer(dly, freqs)), dtype=np.complex64) 
+                        for ant, dly in d_fc_rd.items()}
+    rv['gf_firstcal'] = {ant: np.zeros_like(g, dtype=bool) for ant, g in rv['g_firstcal'].items()}
+    
+    # perform logcal and omnical
+    log_sol = rc.logcal(data, sol0=rv['g_firstcal'])
+    make_sol_finite(log_sol)
+    rv['omni_meta'], omni_sol = rc.omnical(data, log_sol, conv_crit=conv_crit, maxiter=maxiter, 
+                                           check_every=check_every, check_after=check_after, gain=gain)
+
+    # update omnical flags and then remove degeneracies
+    rv['g_omnical'], rv['v_omnical'] = get_gains_and_vis_from_sol(omni_sol)
+    rv['gf_omnical'] = {ant: ~np.isfinite(g) for ant, g in rv['g_omnical'].items()}
+    rv['vf_omnical'] = {bl: ~np.isfinite(v) for bl, v in rv['v_omnical'].items()}
+    make_sol_finite(omni_sol)
+    rd_sol = rc.remove_degen(omni_sol, degen_sol=rv['g_firstcal'])
```

Comment:
We do a little bit, but this is a problem that exists in the current pipeline that abscal already fixes. I don't think we need to add extra functionality here.

Reply:
Could we provide a mechanism for providing an alternate degen sol, in case we want to push abscal into here someday?

Reply:
I don't think we know how we'd want to tie abscal in here in the future, but I think we should wait until we know that before we make this edit. If we're bring in abscal, are we still even running firstcal? Maybe we want a way to bypass firstcal? I don't know.

I think this will be easier to do later and risks adding bloat to our functions, docstrings, and unit tests now. To me this falls under the umbrella overly general code that might be harder to maintain later.

Reply:
Moved to Issue #380